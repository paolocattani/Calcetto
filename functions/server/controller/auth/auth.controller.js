"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
require("../../core/env");
const logger_1 = require("../../core/logger");
const middleware_1 = require("../../core/middleware");
const express_1 = require("express");
const auth_manager_1 = require("../../manager/auth.manager");
const common_response_1 = require("../common.response");
const HttpStatusCode_1 = require("../../../src/@common/models/HttpStatusCode");
const cookies_utils_1 = require("./cookies.utils");
const auth_utils_1 = require("./auth.utils");
const events_1 = require("../../events/events");
const router = express_1.Router();
const registrationController = middleware_1.asyncMiddleware((req, res) => __awaiter(void 0, void 0, void 0, function* () {
    logger_1.logger.info('/register start ');
    const registrationInfo = req.body;
    try {
        const errors = auth_manager_1.isValidRegister(registrationInfo);
        if (errors.length !== 0) {
            return common_response_1.failure(res, { label: 'auth:error.generic' }, 'Registration data is not valid', HttpStatusCode_1.HTTPStatusCode.BadRequest, { errors });
        }
        const model = auth_manager_1.parseBody(registrationInfo);
        const user = yield auth_manager_1.findUserByEmailOrUsername(model.username, model.email);
        if (user) {
            return common_response_1.failure(res, { label: 'auth:server.error.already_exists' }, 'Email or Username alrealdy exists.', HttpStatusCode_1.HTTPStatusCode.BadRequest, {
                errors: [{ label: 'auth:server.error.already_exists' }],
            });
        }
        const record = yield auth_manager_1.registerUser(model, registrationInfo.playerRole);
        if (record) {
            cookies_utils_1.addUserCookies(record, res);
            logger_1.logger.info('/register end ');
            return common_response_1.success(res, { label: 'auth:welcome', options: { username: record.name } }, { user: record });
        }
        else {
            throw new Error('Server Error.!');
        }
    }
    catch (error) {
        return common_response_1.serverError('POST auth/register error ! : ', error, res, {
            errors: ["Errore server non previsto. E' stata avviata la procedura di autodistruzione."],
        });
    }
}));
const wrongCredentials = (res) => common_response_1.failure(res, { label: 'auth:server.error.wrong_credential' }, 'Wrong credentials', HttpStatusCode_1.HTTPStatusCode.Unauthorized);
router.use('/', (req, res, next) => middleware_1.controllerLogger(req, next, 'Auth Controller', '/api/v2/auth'));
router.get('/check', middleware_1.withAuth, middleware_1.asyncMiddleware((req, res) => __awaiter(void 0, void 0, void 0, function* () {
    const additional = { user: req.user };
    return common_response_1.success(res, {
        label: 'auth:server.error.wrong_credential',
        options: { username: req.user.name },
    }, additional);
})));
router.get('/logout', middleware_1.withAuth, middleware_1.asyncMiddleware((req, res) => __awaiter(void 0, void 0, void 0, function* () {
    cookies_utils_1.removeUserCookies(res);
    return common_response_1.success(res, { label: 'auth:logout' });
})));
router.post('/register', registrationController);
router.put('/unsubscribe', middleware_1.withAuth, middleware_1.asyncMiddleware((req, res) => __awaiter(void 0, void 0, void 0, function* () {
    try {
        const { user, uuid } = req;
        events_1.unsubscribe(user, uuid);
        return common_response_1.success(res, { label: '' });
    }
    catch (err) {
        return common_response_1.serverError('PUT tournament/unsubscribe error ! : ', err, res);
    }
})));
router.put('/update', middleware_1.withAuth, middleware_1.asyncMiddleware((req, res) => __awaiter(void 0, void 0, void 0, function* () {
    try {
        let model = auth_manager_1.parseBody(req.body);
        logger_1.logger.info('/update : ', model);
        const user = yield auth_manager_1.findUserByEmailAndUsername(model.email, model.username);
        if (!user) {
            return common_response_1.entityNotFound(res);
        }
        yield user.update(model);
        return common_response_1.success(res, { label: 'common:server.updated' }, { user: auth_manager_1.convertEntityToDTO(user) });
    }
    catch (error) {
        return common_response_1.serverError('PUT auth/update error ! : ', error, res);
    }
})));
router.post('/login', middleware_1.asyncMiddleware((req, res) => __awaiter(void 0, void 0, void 0, function* () {
    const { username, password } = req.body;
    return yield loginUserController(res, username, password);
})));
router.delete('/delete', middleware_1.withAuth, middleware_1.asyncMiddleware((req, res) => __awaiter(void 0, void 0, void 0, function* () {
    const { password } = req.body;
    const { email, username } = req.user;
    return yield deleteUserController(res, username, email, password);
})));
router.post('/test/register', middleware_1.withTestAuth, registrationController);
router.delete('/test/delete', middleware_1.withTestAuth, middleware_1.asyncMiddleware((req, res) => __awaiter(void 0, void 0, void 0, function* () {
    const { password, email, username } = req.body;
    return yield deleteUserController(res, username, email, password);
})));
router.post('/test/login', middleware_1.withTestAuth, middleware_1.asyncMiddleware((req, res) => __awaiter(void 0, void 0, void 0, function* () {
    const { username, password } = req.body;
    return yield loginUserController(res, username, password);
})));
router.get('/test/user', middleware_1.withTestAuth, middleware_1.asyncMiddleware((req, res) => __awaiter(void 0, void 0, void 0, function* () {
    const { username, email } = req.query;
    const user = yield auth_manager_1.findUserByEmailAndUsername(email, username);
    if (user) {
        const userDTO = auth_manager_1.convertEntityToDTO(user);
        return common_response_1.success(res, { label: 'auth:welcome', options: { username: userDTO.name } }, { user: userDTO });
    }
    else {
        return common_response_1.entityNotFound(res);
    }
})));
const loginUserController = (res, username, password) => __awaiter(void 0, void 0, void 0, function* () {
    try {
        logger_1.logger.info('/login start ');
        if (!username || !password) {
            return common_response_1.missingParameters(res);
        }
        const user = yield auth_manager_1.findUserByEmailOrUsername(username, username);
        if (!user) {
            return common_response_1.entityNotFound(res);
        }
        if (!(yield auth_utils_1.comparePasswords(user.email, password, user.password))) {
            return wrongCredentials(res);
        }
        const userDTO = auth_manager_1.convertEntityToDTO(user);
        cookies_utils_1.addUserCookies(userDTO, res);
        return common_response_1.success(res, { label: 'auth:welcome', options: { username: userDTO.name } }, { user: userDTO });
    }
    catch (error) {
        return common_response_1.serverError('POST auth/login error ! : ', error, res);
    }
});
const deleteUserController = (res, username, email, password) => __awaiter(void 0, void 0, void 0, function* () {
    try {
        logger_1.logger.info('/delete start ');
        const user = yield auth_manager_1.findUserByEmailAndUsername(email, username);
        if (!user) {
            return common_response_1.entityNotFound(res);
        }
        if (!(yield auth_utils_1.comparePasswords(email, password, user.password))) {
            return wrongCredentials(res);
        }
        yield auth_manager_1.deleteUser(user);
        logger_1.logger.info('/delete end ');
        cookies_utils_1.removeUserCookies(res);
        return common_response_1.success(res, { label: 'auth:server.deleted', options: { username: user.name } });
    }
    catch (error) {
        return common_response_1.serverError('DELETE auth/delete error ! : ', error, res);
    }
});
exports.default = router;
