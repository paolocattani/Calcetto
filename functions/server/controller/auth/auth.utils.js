"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.safeVerifyToken = exports.comparePasswords = exports.generatePassword = exports.TOKEN_SECRET = void 0;
const bcryptjs_1 = __importDefault(require("bcryptjs"));
const jsonwebtoken_1 = __importDefault(require("jsonwebtoken"));
const DEFAULT_HASH = 'dummy$Hash';
exports.TOKEN_SECRET = process.env.SERVER_HASH || DEFAULT_HASH;
const generateHashSecret = (email, password) => email + exports.TOKEN_SECRET + password;
const generatePassword = (email, password) => bcryptjs_1.default.hash(generateHashSecret(email, password), 10);
exports.generatePassword = generatePassword;
const comparePasswords = (email, password, hash) => bcryptjs_1.default.compare(generateHashSecret(email, password), hash);
exports.comparePasswords = comparePasswords;
const safeVerifyToken = (token) => {
    let decoded = null;
    try {
        decoded = jsonwebtoken_1.default.verify(token, exports.TOKEN_SECRET);
        return [decoded, true];
    }
    catch (error) {
        return [null, false];
    }
};
exports.safeVerifyToken = safeVerifyToken;
