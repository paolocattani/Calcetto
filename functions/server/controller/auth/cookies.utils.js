"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.removeUserCookies = exports.addUserCookies = exports.getCookies = void 0;
const uuid_1 = require("uuid");
const jsonwebtoken_1 = __importDefault(require("jsonwebtoken"));
const debug_1 = require("../../core/debug");
const auth_utils_1 = require("./auth.utils");
const SESSION_TOKEN = 'session_id';
const USER_UUID = 'user_id';
const DEFAULT_TOKEN_EXPIRATION = '8h';
const TOKEN_EXPIRATION = process.env.SERVER_TOKEN_EXPIRES_IN || DEFAULT_TOKEN_EXPIRATION;
const generateUuid = (user) => uuid_1.v5(user.id.toString(), '48990e20-7e42-4f0c-97e0-c64456d5bc71');
const generateToken = (value) => jsonwebtoken_1.default.sign(Object.assign({}, value), auth_utils_1.TOKEN_SECRET, {
    expiresIn: TOKEN_EXPIRATION,
    algorithm: 'HS256',
});
const getCookies = (req) => [getToken(req), getUuid(req)];
exports.getCookies = getCookies;
const getToken = (req) => req.signedCookies[SESSION_TOKEN];
const getUuid = (req) => req.signedCookies[USER_UUID];
const cookiesOption_old = Object.assign({ maxAge: 8 * 60 * 60 * 1000, httpOnly: true, signed: true }, (debug_1.isProductionMode()
    ? {
        secure: true,
        sameSite: 'none',
    }
    : {
        httpOnly: true,
    }));
const cookiesOption = {
    maxAge: 8 * 60 * 60 * 1000,
    httpOnly: true,
    signed: true,
    secure: debug_1.isProductionMode(),
    sameSite: debug_1.isProductionMode() ? 'strict' : 'lax',
};
const addUserCookies = (user, res) => {
    res.cookie(USER_UUID, generateUuid(user), cookiesOption);
    res.cookie(SESSION_TOKEN, generateToken(user), cookiesOption);
};
exports.addUserCookies = addUserCookies;
const removeUserCookies = (res) => {
    res.clearCookie(USER_UUID, cookiesOption);
    res.clearCookie(SESSION_TOKEN, cookiesOption);
};
exports.removeUserCookies = removeUserCookies;
