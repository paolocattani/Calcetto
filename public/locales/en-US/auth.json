{
  "name": "Name",
  "surname": "Surname",
  "username": "Username",
  "usernameEmail": "Username or Email",
  "mobile": "Mobile",
  "birthday": "Birthday",
  "player": "Are you a player? Pick your role here !.",
  "register": "Register",
  "delete": "Delete user",
  "manage": "Manage user",
  "welcome": "Welcome {{username}}",
  "login": "Login",
  "logout": "See you...",
  "expired_alert": "Your session expired. You'll be redirect to login page...",
  "expired": "Your session expired....",
  "cookies": {
    "accept": "Accept",
    "info": "We use cookies to allow you to browse this site and no information is collected."
  },
  "email": {
    "email": "Email",
    "confirm": "Confirm email"
  },
  "password": {
    "password": "Password",
    "confirm": "Confirm password",
    "criteria": {
      "title": "Password must match following criteria",
      "c1": "At least 1 numeric lowercase char.",
      "c2": "At least 1 numeric uppercase char.",
      "c3": "At least 1 numeric char.",
      "c4": "Lenght between 8 and 16."
    }
  },
  "error": {
    "generic": "Fix errors to register",
    "username": "Username is mandatory",
    "name": "Name is mandatory",
    "surname": "Surname is mandatory",
    "email": {
      "address": "Email is mandatory",
      "confirm": "Email confirm is mandatory",
      "validation": "Provided email is not valid",
      "match": "Emails does not match"
    },
    "password": {
      "password": "Password is mandatory",
      "confirm": "Password confirm is mandatory",
      "validation": "Provided password does not match criteria",
      "match": "Provided passwords does not match"
    }
  },
  "server": {
    "deleted": "User {{username}} deleted",
    "error": {
      "wrong_credential": "Wrong crediantial",
      "already_exists": "Email or Username already exists."
    }
  }
}