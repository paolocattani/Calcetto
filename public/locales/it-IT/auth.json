{
    "usernameEmail":"Username o Email",
    "password":"Password",
    "register":"Registrati",
    "error":{
        "username":"Il campo username è obbligatorio",
        "password":"Il campo password è obbligatorio"
    }
}
