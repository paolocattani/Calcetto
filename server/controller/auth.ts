import '../core/env';
import { logger } from '../core/logger';
import { isDevMode } from '../core/debug';

import { Router, Request, Response, NextFunction } from 'express';
import User from '../model/sequelize/user.model';
import { withAuth, asyncMiddleware } from '../core/middleware';

import {
  findUserByEmail,
  convertEntityToDTO,
  parseBody,
  deleteUser,
  comparePasswords,
  generateToken,
  listAll,
  registerUser,
  findUserByEmailOrUsername,
  checkIfExist,
  findUserByEmailAndUsername
} from '../manager/auth';
const router = Router();

router.use('/', (req, res, next) => {
  if (isDevMode()) logger.info(`Auth Manager : ${req.method} ${req.originalUrl.replace('/api/v1/auth', '')} `);
  next();
});

router.get(
  '/list',
  withAuth,
  // FIXME: fix request type
  asyncMiddleware(async (req: any, res: Response, next: NextFunction) => res.status(200).json(await listAll()))
);

router.get(
  '/',
  withAuth,
  // FIXME: fix request type
  asyncMiddleware(async (req: any, res: Response, next: NextFunction) =>
    res.status(200).json(await findUserByEmail(req.email))
  )
);

router.get(
  '/logout',
  withAuth,
  asyncMiddleware(async (req: any, res: Response, next: NextFunction) => {
    // FIXME: logout
    res.cookie('token', { expires: new Date(Date.now()), httpOnly: true });
    return res.status(200).json({ message: 'logout complete' });
  })
);

router.post(
  '/register',
  asyncMiddleware(async (req: Request, res: Response, next: NextFunction) => {
    const { body } = req;
    let model: User = parseBody(body);
    const user = await checkIfExist(model);
    if (user) return res.status(403).json({ message: 'Email o Username gia utilizzati. ' });
    const record = await registerUser(model, body.playerRole);
    return res.status(200).json(record);
  })
);

router.post(
  '/authenticate',
  asyncMiddleware(async (req: Request, res: Response, next: NextFunction) => {
    const { username, password } = req.body;
    try {
      logger.info('/authenticate : ', username, password);
      if (!username || !password) return res.status(401).json({ error: 'Missing email or password' });

      const user = await findUserByEmailOrUsername(username);

      // utente non trovato
      if (!user) return res.status(401).json({ error: 'Incorrect email or password' });

      if (!(await comparePasswords(user.email, password, user.password)))
        return res.status(401).json({ error: 'Incorrect email or password' });

      res.cookie('token', generateToken(user), { httpOnly: true });
      return res.status(200).json(convertEntityToDTO(user));
    } catch (error) {
      logger.error('/authenticate error : ', error);
      return res.status(500).json({ error: 'Internal error please try again' });
    }
  })
);

router.delete(
  '/',
  withAuth,
  asyncMiddleware(async (req: any, res: Response, next: NextFunction) => {
    const { email, password, username } = req.body;
    try {
      logger.info('/authenticate : ', email, password);
      const user = await findUserByEmailAndUsername(email, username);
      if (!user) return res.status(500).json({ message: 'Utente non trovato' });

      if (!(await comparePasswords(email, password, user.password)))
        return res.status(401).json({ error: 'Incorrect email or password' });

      await deleteUser(user);
      res.cookie('token', { expires: new Date(Date.now()), httpOnly: true });
      return res.status(200).json({ message: 'Utente eliminato' });
    } catch (error) {
      return res.status(500).json({ message: 'Errore durante fase di cancellazione' });
    }
  })
);

router.get('/checkToken', withAuth, (req, res) => res.sendStatus(200));

export default router;
