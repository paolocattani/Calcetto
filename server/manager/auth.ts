// Models/Types
import Player from 'model/sequelize/player.model';
import { PlayerRole } from 'model/sequelize/types';
import { UserDTO } from '../model/dto/user';
import User from '../model/sequelize/user.model';
// Logger utils
import { logProcess } from '../core/logger';
// Password
import jwt from 'jsonwebtoken';
import bcrypt from 'bcrypt';
// managers
import * as playerManager from './player';
import { Op } from 'sequelize';

// Const
const className = 'Authentication Manager : ';

// Password utils
export const generatePassword = async (email: string, password: string) =>
  await bcrypt.hash(generateHashSecret(email, password), 10);

export const comparePasswords = async (email: string, password: string, hash: string): Promise<boolean> =>
  await bcrypt.compare(generateHashSecret(email, password), hash);

export const getSecret = () => (process.env.SERVER_HASH ? process.env.SERVER_HASH : 'dummy$Hash');

export const generateHashSecret = (email: string, password: string) => email + getSecret() + password;

export const generateToken = (value: string | User) =>
  typeof value === 'string'
    ? jwt.sign({ email: value }, getSecret(), { expiresIn: '8h', algorithm: 'HS256' })
    : jwt.sign({ name: value.name, surname: value.surname, role: value.role, email: value.email }, getSecret(), {
        expiresIn: '8h',
        algorithm: 'HS256'
      });

export const listAll = async (): Promise<UserDTO[]> => {
  try {
    logProcess(className + 'listAll', 'start');
    const users = await User.findAll({
      order: [['id', 'DESC']],
      include: [User.associations.player]
    });
    logProcess(className + 'listAll', 'end');
    return users.map(user => convertEntityToDTO(user));
  } catch (error) {
    logProcess(className + 'listAll', ` Error : ${error}`);
    return [];
  }
};

export const deleteUser = async (user: User): Promise<void> => {
  try {
    logProcess(className + 'deleteUser', 'start');
    await user.destroy();
    logProcess(className + 'deleteUser', 'end');
  } catch (error) {
    logProcess(className + 'deleteUser', ` Error : ${error}`);
  }
};

export const registerUser = async (user: User, playerRole?: string): Promise<UserDTO | null> => {
  try {
    logProcess(className + 'registerUser', 'start');
    user.password = await generatePassword(user.email, user.password);
    if (user.name.startsWith('[A]')) {
      user.name = user.name.substring(3);
      user.role = 'Admin';
    } else {
      user.role = 'User';
    }
    const record = await User.create(user);
    // Se è stato assegnato un ruolo allora creo anche il giocatore
    if (playerRole) {
      const model = {
        name: record.name,
        surname: record.surname,
        email: record.email,
        phone: record.phone,
        userId: record.id,
        alias: `${record.name} ${record.surname}`,
        role: playerRole as PlayerRole
      } as Player;
      const player = await playerManager.create(model);
      if (player) await record.update({ playerId: player.id });
    }
    logProcess(className + 'registerUser', 'end');
    return convertEntityToDTO(record);
  } catch (error) {
    logProcess(className + 'registerUser', ` Error : ${error}`);
    return null;
  }
};

// Utils
export function isAdmin(token: string | object): boolean {
  let isAdmin: boolean = false;
  if (token && typeof token === 'string') {
    const decodedUser = jwt.verify(token, getSecret()) as User;
    if (decodedUser.role === 'Admin') isAdmin = true;
  }
  return isAdmin;
}

export async function getUserFromToken(token: string | object): Promise<User | null> {
  if (token && typeof token === 'string') {
    const decodedUser = jwt.verify(token, getSecret()) as User;
    return decodedUser.email ? await findUserByEmail(decodedUser.email) : decodedUser;
  } else return null;
}

export async function findUserByEmail(email: string) {
  try {
    logProcess(className + 'findUserByEmail', '');
    return await User.findOne({ where: { email } });
  } catch (error) {
    logProcess(className + 'findUserByEmail', ` Error : ${error}`);
    return null;
  }
}

export async function findUserByEmailOrUsername(username: string) {
  try {
    logProcess(className + 'findUserByEmailOrUsername', '');
    return await User.findOne({ where: { [Op.or]: [{ email: username }, { username }] } });
  } catch (error) {
    logProcess(className + 'findUserByEmailOrUsername', ` Error : ${error}`);
    return null;
  }
}

export async function checkIfExist(user: User) {
  try {
    logProcess(className + 'checkIfExist', '');
    return await User.findOne({ where: { [Op.or]: [{ email: user.email }, { username: user.username }] } });
  } catch (error) {
    logProcess(className + 'checkIfExist', ` Error : ${error}`);
    return null;
  }
}

/**
 * Converte l'entity in un DTO da passare al FE.
 * @param user  User entity
 */
export const convertEntityToDTO = (user: User): UserDTO => ({
  id: user.id,
  username: user.username,
  name: user.name,
  surname: user.surname,
  email: user.email,
  phone: user.phone || '',
  birthday: user.birthday || null,
  label: user.label,
  role: user.role
});

/**
 * Si potrebbe usare lo spread {...body} ma essendo dati che arrivano dall'utente
 * preferisco impostare le proprietà manualmente.
 *
 * @param body : Request body
 */
export const parseBody = (body: any) =>
  ({
    id: body.id || null,
    username: body.username,
    name: body.name,
    surname: body.surname,
    password: body.password,
    email: body.email,
    phone: body.phone,
    birthday: body.birthday,
    role: body.role
  } as User);
