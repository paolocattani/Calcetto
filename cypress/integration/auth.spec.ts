// enables intelligent code completion for Cypress commands
// https://on.cypress.io/intelligent-code-completion


import {LandingPage, RegistrationProps} from "../pages/landing.page";
import { AuthAction } from '../../src/redux/actions/auth.action';

const {users:{admin,user}} = require('../fixtures/auth.fixture.json')

describe('Authentication Test', () => {
	before(() => {
		Cypress.Cookies.debug(true);
	});

	beforeEach(function() {
		//Cypress.Cookies.preserveOnce('session_id')
  });

	function afterLogin (landingPage:LandingPage,user:{username:string}){
		landingPage.getHeaderUsername().should('be.visible').and('contain',user.username);
		cy.getCookies()
			.should('have.length', 1)
			.then((cookies) => {
				const session_id = cookies[0];
				expect(session_id).to.have.property('name', 'session_id');
				expect(session_id).to.have.property('domain','localhost');
				expect(session_id).to.have.property('httpOnly');
				expect(session_id).to.have.property('path');
				expect(session_id).to.have.property('secure',false);
			});

		// https://www.cypress.io/blog/2018/11/14/testing-redux-store/
		// Dispatch action   cy.window().its('store').invoke('dispatch', AuthAction.login.... )
		cy.window().its('store').invoke('getState').toMatchSnapshot();


	}

  describe('Registration process', function () {

		it('Should register Admin', ()=> {
			const landingPage = new LandingPage();
			landingPage.forceDeleteUser(admin.email,admin.username,admin.password);
			// https://glebbahmutov.com/blog/keep-passwords-secret-in-e2e-tests/
			landingPage.register(admin,true);
			afterLogin(landingPage,admin);
		});

    it('Should register User', ()=> {
			const landingPage = new LandingPage();
			landingPage.forceDeleteUser(user.email,user.username,user.password);
			landingPage.register(user,false);
			afterLogin(landingPage,user);
    });
  });

  describe('Login process', ()=> {
    it('Should login as Admin', ()=> {
				const landingPage = new LandingPage();
				landingPage.login(admin);
				afterLogin(landingPage,admin);
				landingPage.forceLogout();
		});

    it('Should login as User', ()=> {
				const landingPage = new LandingPage();
				landingPage.login(user);
				afterLogin(landingPage,user);
				landingPage.forceLogout();
    });
  });

  // more examples
  //
  // https://github.com/cypress-io/cypress-example-todomvc
  // https://github.com/cypress-io/cypress-example-kitchensink
  // https://on.cypress.io/writing-your-first-test
});
