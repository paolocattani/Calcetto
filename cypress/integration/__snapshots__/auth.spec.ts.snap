exports[`Authentication Test > Registration process > Should register Admin #0`] =
{
  "_persist": {
    "rehydrated": true,
    "version": -1
  },
  "authState": {
    "isAdmin": true,
    "isAuthenticated": true,
    "isLoading": false,
    "user": {
      "birthday": "2020-11-30T17:46:15.319Z",
      "email": "admin@gmail.com",
      "id": 1,
      "label": "Admin Admin",
      "name": "Admin",
      "phone": "3672948995",
      "role": "Admin",
      "surname": "Admin",
      "username": "Admin"
    }
  },
  "pairState": {
    "isLoading": false,
    "isSaving": false
  },
  "playerState": {
    "isLoading": false,
    "isSaving": false,
    "playersList": []
  },
  "stage1State": {
    "isLoading": false,
    "needRefresh": false,
    "selectedPairs": [
      {
        "alias": "-",
        "id": null,
        "paid1": false,
        "paid2": false,
        "placement": 0,
        "player1": {
          "alias": "",
          "editable": false,
          "email": "",
          "id": null,
          "label": "",
          "match_played": 0,
          "match_won": 0,
          "name": "",
          "phone": "",
          "role": "Portiere",
          "rowNumber": 0,
          "surname": "",
          "total_score": 0
        },
        "player2": {
          "alias": "",
          "editable": false,
          "email": "",
          "id": null,
          "label": "",
          "match_played": 0,
          "match_won": 0,
          "name": "",
          "phone": "",
          "role": "Portiere",
          "rowNumber": 0,
          "surname": "",
          "total_score": 0
        },
        "rowNumber": 0,
        "stage1Name": "",
        "tournamentId": 0
      }
    ],
    "stages": []
  },
  "stage2State": {
    "isLoading": false
  },
  "tournamentState": {
    "isLoading": true,
    "tournament": null,
    "tournamentsList": []
  }
};

exports[`Authentication Test > Registration process > Should register User #0`] =
{
  "_persist": {
    "rehydrated": true,
    "version": -1
  },
  "authState": {
    "isAdmin": false,
    "isAuthenticated": true,
    "isLoading": false,
    "user": {
      "birthday": "2020-11-30T17:46:22.097Z",
      "email": "user@gmail.com",
      "id": 2,
      "label": "User User",
      "name": "User",
      "phone": "3672948006",
      "role": "User",
      "surname": "User",
      "username": "User"
    }
  },
  "pairState": {
    "isLoading": false,
    "isSaving": false
  },
  "playerState": {
    "isLoading": false,
    "isSaving": false,
    "playersList": []
  },
  "stage1State": {
    "isLoading": false,
    "needRefresh": false,
    "selectedPairs": [
      {
        "alias": "-",
        "id": null,
        "paid1": false,
        "paid2": false,
        "placement": 0,
        "player1": {
          "alias": "",
          "editable": false,
          "email": "",
          "id": null,
          "label": "",
          "match_played": 0,
          "match_won": 0,
          "name": "",
          "phone": "",
          "role": "Portiere",
          "rowNumber": 0,
          "surname": "",
          "total_score": 0
        },
        "player2": {
          "alias": "",
          "editable": false,
          "email": "",
          "id": null,
          "label": "",
          "match_played": 0,
          "match_won": 0,
          "name": "",
          "phone": "",
          "role": "Portiere",
          "rowNumber": 0,
          "surname": "",
          "total_score": 0
        },
        "rowNumber": 0,
        "stage1Name": "",
        "tournamentId": 0
      }
    ],
    "stages": []
  },
  "stage2State": {
    "isLoading": false
  },
  "tournamentState": {
    "isLoading": true,
    "tournament": null,
    "tournamentsList": []
  }
};

exports[`Authentication Test > Login process > Should login as Admin #0`] =
{
  "_persist": {
    "rehydrated": true,
    "version": -1
  },
  "authState": {
    "isAdmin": true,
    "isAuthenticated": true,
    "isLoading": false,
    "user": {
      "birthday": "2020-11-30T17:47:20.126Z",
      "email": "admin@gmail.com",
      "id": 3,
      "label": "Admin Admin",
      "name": "Admin",
      "phone": "3672948995",
      "role": "Admin",
      "surname": "Admin",
      "username": "Admin"
    }
  },
  "pairState": {
    "isLoading": false,
    "isSaving": false
  },
  "playerState": {
    "isLoading": false,
    "isSaving": false,
    "playersList": []
  },
  "stage1State": {
    "isLoading": false,
    "needRefresh": false,
    "selectedPairs": [
      {
        "alias": "-",
        "id": null,
        "paid1": false,
        "paid2": false,
        "placement": 0,
        "player1": {
          "alias": "",
          "editable": false,
          "email": "",
          "id": null,
          "label": "",
          "match_played": 0,
          "match_won": 0,
          "name": "",
          "phone": "",
          "role": "Portiere",
          "rowNumber": 0,
          "surname": "",
          "total_score": 0
        },
        "player2": {
          "alias": "",
          "editable": false,
          "email": "",
          "id": null,
          "label": "",
          "match_played": 0,
          "match_won": 0,
          "name": "",
          "phone": "",
          "role": "Portiere",
          "rowNumber": 0,
          "surname": "",
          "total_score": 0
        },
        "rowNumber": 0,
        "stage1Name": "",
        "tournamentId": 0
      }
    ],
    "stages": []
  },
  "stage2State": {
    "isLoading": false
  },
  "tournamentState": {
    "isLoading": true,
    "tournament": null,
    "tournamentsList": []
  }
};

exports[`Authentication Test > Login process > Should login as User #0`] =
{
  "_persist": {
    "rehydrated": true,
    "version": -1
  },
  "authState": {
    "isAdmin": false,
    "isAuthenticated": true,
    "isLoading": false,
    "user": {
      "birthday": "2020-11-30T17:47:25.817Z",
      "email": "user@gmail.com",
      "id": 4,
      "label": "User User",
      "name": "User",
      "phone": "3672948006",
      "role": "User",
      "surname": "User",
      "username": "User"
    }
  },
  "pairState": {
    "isLoading": false,
    "isSaving": false
  },
  "playerState": {
    "isLoading": false,
    "isSaving": false,
    "playersList": []
  },
  "stage1State": {
    "isLoading": false,
    "needRefresh": false,
    "selectedPairs": [
      {
        "alias": "-",
        "id": null,
        "paid1": false,
        "paid2": false,
        "placement": 0,
        "player1": {
          "alias": "",
          "editable": false,
          "email": "",
          "id": null,
          "label": "",
          "match_played": 0,
          "match_won": 0,
          "name": "",
          "phone": "",
          "role": "Portiere",
          "rowNumber": 0,
          "surname": "",
          "total_score": 0
        },
        "player2": {
          "alias": "",
          "editable": false,
          "email": "",
          "id": null,
          "label": "",
          "match_played": 0,
          "match_won": 0,
          "name": "",
          "phone": "",
          "role": "Portiere",
          "rowNumber": 0,
          "surname": "",
          "total_score": 0
        },
        "rowNumber": 0,
        "stage1Name": "",
        "tournamentId": 0
      }
    ],
    "stages": []
  },
  "stage2State": {
    "isLoading": false
  },
  "tournamentState": {
    "isLoading": true,
    "tournament": null,
    "tournamentsList": []
  }
};
